src/discof/repair/fd_repair_tile.c

```diff
@@ -247,40 +278,100 @@ loose_footprint( fd_topo_tile_t const * tile FD_PARAM_UNUSED ) {
+
+  fd_repair_pending_sign_t * pending = fd_repair_pending_sign_pool_ele_acquire( ctx->pending_sign_pool );
+  if( FD_UNLIKELY( !pending ) ) return NULL;
+
+  pending->msg    = *msg;
+  pending->nonce  = pending->msg.shred.nonce;
+  pending->buflen = fd_repair_sz( msg );
+
+  fd_repair_pending_sign_map_ele_insert( ctx->pending_sign_map, pending, ctx->pending_sign_pool );
+  return pending;
+}
+
+fd_repair_pending_sign_t *
+pending_sign_request_query( ctx_t * ctx,
+                            ulong   nonce ) {
+  return fd_repair_pending_sign_map_ele_query( ctx->pending_sign_map, &nonce, NULL, ctx->pending_sign_pool );
+}
+
+int
+pending_sign_request_remove( ctx_t * ctx,
+                             ulong   nonce  ) {
+  fd_repair_pending_sign_t * pending = fd_repair_pending_sign_map_ele_query( ctx->pending_sign_map, &nonce, NULL, ctx->pending_sign_pool );
+  if( FD_UNLIKELY( !pending ) ) return -1;
+  fd_repair_pending_sign_map_ele_remove  ( ctx->pending_sign_map, &nonce, NULL, ctx->pending_sign_pool );
+  fd_repair_pending_sign_pool_ele_release( ctx->pending_sign_pool, pending );
+  return 0;
+}
 
 /* Wrapper for keyguard client sign */
 static void
-repair_signer( void *        signer_ctx,
-               uchar         signature[ static 64 ],
-               uchar const * buffer,
-               ulong         len,
-               int           sign_type ) {
-  fd_repair_tile_ctx_t * ctx = (fd_repair_tile_ctx_t *) signer_ctx;
+repair_signer_sync( ctx_t *       ctx,
+                    uchar         signature[ static 64 ],
+                    uchar const * buffer,
+                    ulong         len,
+                    int           sign_type ) {
   fd_keyguard_client_sign( ctx->keyguard_client, signature, buffer, len, sign_type );
 }
 
 /* Wrapper for publishing to the sign tile*/
 static void
-repair_signer_async( void *        signer_ctx,
-                     ulong         nonce,
-                     uchar const * buffer,
-                     ulong         len,
-                     int           sign_type,
-                     fd_repair_out_ctx_t * sign_out) {
-  fd_repair_tile_ctx_t * ctx = (fd_repair_tile_ctx_t *) signer_ctx;
+repair_signer_async( ctx_t *                    ctx,
+                     fd_repair_pending_sign_t * pending,
+                     int                        sign_type,
+                     out_ctx_t *                sign_out) {
+  ulong   nonce  = pending->nonce;
+  uchar * buffer = pending->buf;
+  ulong   len    = pending->buflen;
+
+   /* At this point buffer contains
+
+     [ discriminant ] [ empty sig ] [ payload ]
+     ^                ^             ^
+     0                4             68
+
+     https://github.com/solana-labs/solana/blob/master/core/src/repair/serve_repair.rs#L1258
+
+     We want to sign over
+     [ discriminant ] [ payload ]
+      ^                ^
+      buffer           buffer+4
+   */
+
+  fd_memcpy( buffer+64, buffer, 4 );
+  buffer += 64UL;
+  len    -= 64UL;
```

Comment:
i would re-implement and document this more clearly inside `fd_repair.c` (think you can take parts of my impl for that?)